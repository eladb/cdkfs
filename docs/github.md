# GitHub

By default, many projects are initialized with `GitHub` component to enabled GitHub as the default provider for CI/CD workflows. See https://docs.github.com/en/actions for more information.

The use of GitHub (and generating corresponding files in `.github`) can be disabled by specifying `github: false` in your project options.

## Workflows

TODO

See the `GitHub`, `GithubWorkflow`, and `Job` types in the [API reference](./api/API.md) for currently available APIs.

Example code of creating a GitHub workflow: <https://github.com/projen/projen/blob/65b4194c163f47ba4842981b0c92dbe516be787b/src/github/auto-approve.ts#L67-L105>

### Stale workflow

A "stale" workflow can be added which will automatically close issues or pull
requests on your GitHub repository after time has passed without the issue
seeing any comments or updates. You can enable as shown below:

```ts
// or PythonProject, etc.
new typescript.TypeScriptProject({
  stale: true,
  staleOptions: {
    issues: {
      closeMessage: "closing pull request",
      staleLabel: "I-AM-STALE",
      daysBeforeStale: 180,
    },
  }
})
```

Check the API reference for a list of all available options.

When enabled, by default issues with no activity with will be marked as stale
after 60 days and closed within 7 days, and pull requests with no activity will
be marked as stale after 14 days and closed within 2 days.
