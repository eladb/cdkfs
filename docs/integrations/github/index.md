---
sidebar_position: 1
---

# GitHub Integration


This is just a placeholder doc. You can contribute to this project! Please see the [Contributing Guide](https://github.com/projen/projen/blob/main/docusaurus/README.md) for more information.
