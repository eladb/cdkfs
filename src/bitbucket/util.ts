

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}