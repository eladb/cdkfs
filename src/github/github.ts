import { Component } from '../component';
import { FileBase } from '../file';
import { Project } from '../project';
import { Dependabot, DependabotOptions } from './dependabot';
import { Mergify } from './mergify';
import { PullRequestTemplate } from './pr-template';
import { GithubWorkflow } from './workflows';

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;
}

export class GitHub extends Component {
  /**
   * The `Mergify` configured on this repository. This is `undefined` if Mergify
   * was not enabled when creating the repository.
   */
  public readonly mergify?: Mergify;

  /**
   * The .gitattributes file for this repository.
   */
  private readonly gitattributes: GitattributesFile;

  public constructor(project: Project, options: GitHubOptions) {
    super(project);

    this.gitattributes = new GitattributesFile(project);

    // Contents  of the .projen/ directory are generated by projen
    this.annotateGenerated('/.projen/**');
    // The .gitattributes file itself is generated
    this.annotateGenerated(`/${this.gitattributes.path}`);

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this);
    }
  }

  public addWorkflow(name: string) {
    return new GithubWorkflow(this, name);
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }

  /**
   * Marks the provided file(s) as being generated. This is achieved using the
   * github-linguist attributes. Generated files do not count against the
   * repository statistics and language breakdown.
   *
   * @param glob the glob pattern to match (could be a file path).
   *
   * @see https://github.com/github/linguist/blob/master/docs/overrides.md
   */
  public annotateGenerated(glob: string): void {
    this.gitattributes.addAttributes(glob, 'linguist-generated');
  }
}

class GitattributesFile extends FileBase {
  private readonly attributes = new Map<string, Set<string>>();

  public constructor(project: Project) {
    super(project, '.gitattributes');
  }

  public addAttributes(glob: string, ...attributes: readonly string[]) {
    if (!this.attributes.has(glob)) {
      this.attributes.set(glob, new Set());
    }
    const set = this.attributes.get(glob)!;
    for (const attribute of attributes) {
      set.add(attribute);
    }
  }

  protected synthesizeContent(): string | undefined {
    // We can assume the file map is never empty.
    const entries = Array.from(this.attributes.entries())
      .sort(([l], [r]) => l.localeCompare(r));

    const maxLen = Math.max(...entries.map(([glob]) => glob.length));
    return [
      `# ${FileBase.PROJEN_MARKER}`,
      '',
      ...entries.map(([name, attributes]) => `${name.padEnd(maxLen, ' ')}\t${Array.from(attributes).join(' ')}`),
    ].join('\n');
  }
}
