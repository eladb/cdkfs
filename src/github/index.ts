export * from './github';

export * from './dependabot';
export * from './workflows';
export * from './mergify';
export * from './pr-template';
export * from './mergify';
export * from './auto-merge';
export * from './auto-approve';
export * from './stale';

export * from './task-workflow';

export * as workflows from './workflows-model';
