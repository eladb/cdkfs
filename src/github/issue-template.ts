import * as path from "path";
import { IConstruct } from "constructs";
import * as YAML from "yaml";
import { Component } from "../component";
import { TextFile } from "../textfile";
import { YamlFile } from "../yaml";

export interface ContactLinks {
  /**
   * The name of the contact link.
   */
  readonly name: string;

  /**
   * The URL of the contact link.
   */
  readonly url: string;

  /**
   * A brief description of the contact link.
   */
  readonly about: string;
}

/**
 * Options for configuring issue templates in a repository.
 *
 * @interface IssueTemplateConfigOptions
 * @see {@link https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#configuring-the-template-chooser}
 */
export interface IssueTemplateConfigOptions {
  /**
   * Indicates whether blank issues (issues without a template) are allowed.
   *
   * @type {boolean} [blankIssuesEnabled=false]
   */
  readonly blankIssuesEnabled?: boolean;

  /**
   * An array of contact links to display in the issue template chooser.
   *
   */
  readonly contactLinks?: ContactLinks[];
}

/**
 * Options for `IssueTemplate`.
 */
export interface IssueTemplateOptions {
  /**
   * An array of issue template file names and their contents.
   * The file names should include the extension (.md or .yml).
   */
  readonly templates: { [fileName: string]: string };

  /**
   * The path within the repository where the issue templates should be created.
   * Defaults to '.github/ISSUE_TEMPLATE'.
   */
  readonly templatePath?: string;

  /**
   * Configuration options for the config.yml file.
   * Only applicable if includeConfig is true.
   */
  readonly configOptions?: IssueTemplateConfigOptions;
}

/**
 * Generates Issue Templates for GitHub repositories
 * see  {@link https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests}
 */
export class IssueTemplate extends Component {
  constructor(scope: IConstruct, options: IssueTemplateOptions) {
    super(scope);
    const project = this.project;

    // Create issue template files
    this.addTemplates(options);

    // Create config.yml file if requested
    if (options.configOptions) {
      const blankIssues = options.configOptions?.blankIssuesEnabled;
      const contactLinks = options.configOptions?.contactLinks;

      new YamlFile(project, ".github/config.yml", {
        obj: {
          blank_issues_enabled: blankIssues,
          contact_links: contactLinks,
        },
        marker: false,
        committed: true,
      });
    }
  }

  /**
   * Adds issue templates to the project.
   *
   * @param options - The options for adding issue templates.
   *
   * @throws {Error} If an unsupported file extension is encountered.
   */
  public addTemplates(options: IssueTemplateOptions) {
    const templatePath = ".github/ISSUE_TEMPLATE";

    for (const [fileName, content] of Object.entries(options.templates)) {
      const filePath = path.join(templatePath, fileName);
      const fileExtension = path.extname(fileName).toLowerCase();
      switch (fileExtension) {
        case ".md":
          new TextFile(this.project, filePath, {
            lines: [content],
            marker: false,
            committed: true,
          });
          break;
        case ".yml":
          new YamlFile(this.project, filePath, {
            obj: YAML.parse(content),
            marker: false,
            committed: true,
          });
          break;
        default:
          throw new Error(`Unsupported file extension: ${fileExtension}`);
      }
    }
  }
}
