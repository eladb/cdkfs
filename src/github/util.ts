export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
