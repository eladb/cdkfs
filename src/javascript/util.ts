import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
