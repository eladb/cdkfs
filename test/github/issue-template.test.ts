import * as YAML from "yaml";
import { IssueTemplate } from "../../src/github/issue-template";
import { synthSnapshot, TestProject } from "../util";

describe("issue-template", () => {
  test("default project behavior", () => {
    const project = new TestProject();
    expect(
      synthSnapshot(project)[".github/ISSUE_TEMPLATE/bug-report.md"]
    ).toBeUndefined();
  });

  test("include config", () => {
    const project = new TestProject();
    new IssueTemplate(project, {
      templates: {
        "bug-report.md": "<!-- Bug report template content -->",
      },
      configOptions: {
        blankIssuesEnabled: false,
        contactLinks: [
          {
            name: "Support",
            url: "https://example.com/support",
            about: "Get support for this project",
          },
        ],
      },
    });
    const snapshot = synthSnapshot(project);
    const issueTemplateConfig = snapshot[".github/config.yml"];
    expect(issueTemplateConfig).toContain("blank_issues_enabled: false");
    expect(issueTemplateConfig).toContain("contact_links");
    expect(issueTemplateConfig).toMatchSnapshot();
  });

  test("includes bug-report.md", () => {
    const project = new TestProject();
    new IssueTemplate(project, {
      templates: {
        "bug-report.md": "<!-- Bug report template content -->",
      },
    });

    expect(
      synthSnapshot(project)[".github/ISSUE_TEMPLATE/bug-report.md"]
    ).toMatchSnapshot();
  });

  test("includes feature-request.yml", () => {
    const project = new TestProject();
    const data = {
      name: "Feature Request",
      description: "Suggest a new feature for the project",
      title: "[FEATURE] ",
      body: [
        {
          type: "markdown",
          attributes: {
            value:
              "Thanks for taking the time to fill out this feature request",
          },
        },
      ],
    };

    new IssueTemplate(project, {
      templates: {
        "feature-request.yml": YAML.stringify(data),
      },
    });

    expect(
      synthSnapshot(project)[".github/ISSUE_TEMPLATE/feature-request.yml"]
    ).toMatchSnapshot();
  });

  test("throws error if file extension is not supported", () => {
    const project = new TestProject();
    // Define a function that creates an IssueTemplate with invalid file extension
    const createIssueTemplateWithInvalidExtension = () => {
      new IssueTemplate(project, {
        templates: {
          "feature-request.txt": "<!-- some unsupported text -->",
        },
      });
    };

    expect(createIssueTemplateWithInvalidExtension).toThrow();
  });
});
