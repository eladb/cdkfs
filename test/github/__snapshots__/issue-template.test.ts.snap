// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue-template include config 1`] = `
"
blank_issues_enabled: false
contact_links:
  - name: Support
    url: https://example.com/support
    about: Get support for this project
"
`;

exports[`issue-template includes bug-report.md 1`] = `"<!-- Bug report template content -->"`;

exports[`issue-template includes feature-request.yml 1`] = `
"
name: Feature Request
description: Suggest a new feature for the project
title: "[FEATURE] "
body:
  - type: markdown
    attributes:
      value: Thanks for taking the time to fill out this feature request
"
`;
